src/uu/wc/src/wc.rs

```diff
@@ -113,68 +125,216 @@ pub mod options {
     pub static TOTAL: &str = "total";
     pub static WORDS: &str = "words";
 }
-
 static ARG_FILES: &str = "files";
 static STDIN_REPR: &str = "-";
 
+static QS_ESCAPE: &QuotingStyle = &QuotingStyle::Shell {
+    escape: true,
+    always_quote: false,
+    show_control: false,
+};
+static QS_QUOTE_ESCAPE: &QuotingStyle = &QuotingStyle::Shell {
+    escape: true,
+    always_quote: true,
+    show_control: false,
+};
+
+/// Supported inputs.
+#[derive(Debug)]
+enum Inputs<'a> {
+    /// Default Standard input, i.e. no arguments.
+    Stdin,
+    /// Files; "-" means stdin, possibly multiple times!
+    Paths(Vec<Input<'a>>),
+    /// --files0-from; "-" means stdin.
+    Files0From(Input<'a>),
+}
+
+impl<'a> Inputs<'a> {
+    fn new(matches: &'a ArgMatches) -> UResult<Self> {
+        let arg_files = matches.get_many::<OsString>(ARG_FILES);
+        let files0_from = matches.get_one::<OsString>(options::FILES0_FROM);
+
+        match (arg_files, files0_from) {
+            (None, None) => Ok(Self::Stdin),
+            (Some(files), None) => Ok(Self::Paths(files.map(Into::into).collect())),
+            (None, Some(path)) => {
+                // If path is a file, and the file isn't too large, we'll load it ahead
+                // of time. Every path within the file will have its length checked to
+                // hopefully better align the output columns.
+                let input = Input::from(path);
+                match input.try_as_files0()? {
+                    Some(paths) => Ok(Self::Paths(paths)),
+                    None => Ok(Self::Files0From(input)),
+                }
+            }
+            (Some(_), Some(_)) => Err(WcError::FilesDisabled.into()),
+        }
+    }
+
+    // Creates an iterator which yields values borrowed from the command line arguments.
+    // Returns an error if the file specified in --files0-from cannot be opened.
+    fn try_iter(
+        &'a self,
+        settings: &'a Settings<'a>,
```

Comment:
The structure of functions here is what gets you into trouble with a lot of references and the `Cow` in `Input`. What if you make `Input` always a reference and then structure it like this:
```rust
fn count_inputs(inputs: Inputs) -> ... {
    match inputs {
        Inputs::Stdin => count_input(Input::Stdin(StdinKind::Implicit)),
        Inputs::Paths(paths) => /* call count_input for each path */,
        Self::Files0From(path) => /* call count_input for each line */
}
```
Then you don't have to pass `Iterators` around which makes lifetimes a lot easier.

Reply:
The iterator returned by `try_iter` takes care of so much, though.  It also helps keep the flow of the program consistent with what was happening before.  It used to be that it would only iterate through a `[Input]`, then we couldn't handle streams.  I wanted to be able to handle streams from `--files0-from` in the same manner as a list of command line arguments, and it has also made it easy to consistently handle errors like empty file names.

I should have mentioned already, and I'm embarrassed I still haven't gotten to it yet, but it was my intent to write some more tests to capture the new cases in which this change makes this `wc` act like GNU's.  The formatting of the errors when empty paths are found in either a `--files0-from` or the command line is one "feature" we get with this.